staging/src/k8s.io/apiserver/pkg/registry/generic/testing/tester.go

```diff
@@ -193,10 +222,95 @@ func (t *Tester) emitObject(obj runtime.Object, action string) error {
 		if err != nil {
 			return err
 		}
-		_, _, err = t.storage.Delete(ctx, accessor.GetName(), nil)
+		_, _, err = t.genericStorage.Delete(ctx, accessor.GetName(), nil)
 	default:
 		err = fmt.Errorf("unexpected action: %v", action)
 	}
 
 	return err
 }
+
+// shallowCopyStorageInterceptor intercepts Get and List by:
+// - passing fresh runtime.Objects to the underlying etcd storage
+// - shallow copying those filled runtime.Objects to the objects provided by the caller
+//   (in the case of the List this is done for each list item).
+// The interceptee is the last object passed down to the etcd storage.
+type shallowCopyStorageInterceptor struct {
+	storage.Interface
+
+	lock        sync.RWMutex
+	interceptee runtime.Object
+}
+
+// lastReadStorageObject returns the last shared object returned from the underlying storage.
+// This is shallow copied to the consume of this storage.
+func (s *shallowCopyStorageInterceptor) lastReadStorageObject() runtime.Object {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	return s.interceptee
+}
+
+// Get unmarshals json found at key into objPtr. On a not found error, will either
+// return a zero object of the requested type, or an error, depending on ignoreNotFound.
+// Treats empty responses and nil response nodes exactly like a not found error.
+// The returned contents may be delayed, but it is guaranteed that they will
+// be have at least 'resourceVersion'.
+func (s *shallowCopyStorageInterceptor) Get(ctx context.Context, key string, resourceVersion string, objPtr runtime.Object, ignoreNotFound bool) error {
```

Comment:
sorry to keep stalling on this... I always get to this point, get concerned about parallel implementations of behavior drifting (one in storage, and one here in test), and get stuck trying to think of an alternative that seems maintainable.